### DISCUSSION TITLE
CS0311 no implicit reference conversion between two inheriting generics

### DISCUSSION THREAD

**Discussion by PlayCreatively** [dotnet/csharplang#6455] [Q&A] [answered]
```c#
gameObject.AddComponent<Renderer>();
```
I'm getting this error for the line above

```
Error	CS0311	The type 'Renderer' cannot be used as type parameter 'T' in the generic type or method 'Extensions.AddComponent<T>(IGameObject)'. There is no implicit reference conversion from 'Renderer' to 'IComponent<IGameObject>'.
```
This is the extension method `AddComponent` that I'm invoking
```c#
public static T AddComponent<T>(this IGameObject gameObject) where T : IComponent<IGameObject>, new()
```
This is the `Renderer` class which inherits `Component<GameObject>`...
```c#
public sealed class Renderer : Component<GameObject>, IHasExtents
```
...which inherits `IComponent<T> where T : IGameObject`
```c#
public abstract class Component<T> : GameComponent, IComponent<T> where T : IGameObject
```

As you can see, `Renderer` should inherit `IComponent<IGameObject>`, so why do I get an error that there is no implicit reference conversion from 'Renderer' to 'IComponent\<IGameObject\>'?

**Comment by HaloFour** [comment, marked as answer]
It's a question of variance.  `Component<GameObject>` is a subtype of `IComponent<GameObject>`, but `IComponent<GameObject>` is not a subtype of `IComponent<IGameObject>`.

**Reply by PlayCreatively** [reply]
Ah, thank you very much! Changing `IComponent<GameObject>` to `IComponent<IGameObject>` did the trick.
But I do find that to be peculiar since `GameObject` implements `IGameObject`, so I would have thought it would be equivalent 🤔

**Reply by Richiban** [reply]
Changing `IComponent<T>` to `IComponent<out T>` should accomplish what you want; you'll be more limited in what members you can put in the interface though.
